packages/api/cms-api/src/page-tree/page-tree-read-api.ts

```diff
@@ -221,6 +226,35 @@ export function createReadApi(
             });
         },
 
+        async getPaginatedNodes(options = {}, where = {}): Promise<[PageTreeNodeInterface[], number]> {
```

Comment:
while it *might* make sense to have a paginated and non-paginated (old for compat) version in the graphql api, it should not be done here, especially not in the (relatively highly optimized) PageTreeReadApi.

Why not extend getNodes (+queryNodes) by pagination?

alternative: access the repository directly in the resolver

Reply:
I have now extended getNodes so that paging is possible. I have also added a getNodesCount method to get the totalCount for paging.

I also had to add em.flush() back in the subscriber, because it doesn't work without it. The last time I tested it, the changes were probably not applied.

Reply:
> I have now extended getNodes so that paging is possible. I have also added a getNodesCount method to get the totalCount for paging.

:+1: well done!

> I also had to add em.flush() back in the subscriber, because it doesn't work without it. The last time I tested it, the changes were probably not applied.

Please analzye why it didn't work (probably related to the time when the subscriber is called)

Reply:
> Please analzye why it didn't work (probably related to the time when the subscriber is called)

I did some digging, this is more complicated than we think. First of all, flush() should not be called in the event listeners: https://mikro-orm.io/docs/events#using-onflush-event

<img width="976" alt="Screenshot 2023-03-21 at 15 06 20" src="https://user-images.githubusercontent.com/48853629/226631514-05b30a88-e026-409d-8323-f4d611fa019a.png">

What we try to achieve should be done in the `onFlush` event by creating a new change set: https://mikro-orm.io/docs/events#using-onflush-event. I've tried to implement it and noticed that it works when saving an already existing page, but doesn't work for a new page. This is due to our attaching of the document to the page tree node. Attaching works the following:

1. Persist the new page
2. Attach it to the page tree node
3. Flush

However, MikroORM performs a flush before attaching automatically. This leads to the case where the newly created page is flushed without having been attached to the page tree node, consequently we can't update the timestamp of the page tree node. The flush in the third step is ignored as far as I see it.

To fix this, I propose to check for changes in both the document entities and the `AttachedDocument` entity.

Reply:
Here's how the fix would look like: 948af38f7fffe01e30323f7bd026fe8248a5da15

Reply:
> First of all, flush() should not be called in the event listeners

:+1:

> However, MikroORM performs a flush before attaching automatically.

really? in one request? this should not happen.

Reply:
> really? in one request? this should not happen.

I believe this is due to `FlushMode.AUTO`, which is the default value:

> FlushMode.AUTO will try to detect changes on the entity we are querying, and flush if there is an overlap

https://mikro-orm.io/docs/unit-of-work#flush-modes

It could also be due to a wrong flush in the page tree module. Might have to analyze more

Reply:
> I believe this is due to FlushMode.AUTO

ah, ok - that makes sense.

maybe you could avoid the check (and query and thus commit) by using upsert?

Reply:
> maybe you could avoid the check (and query and thus commit) by using upsert?

I've tried that, but that doesn't work, as we don't know the `id` of in `AttachedDocument`, causing upsert to fail on the unique constraint for `pageTreeNodeId` and `documentId`.

> Don't check if the document really exists

The only way to remove the second flush is by removing this check. Upsert wouldn't check if the document really exists, because there is no "real" database relation between attached document and the actual document.

How should we proceed here? Ignore the second flush for now?